sriov/sriov.go

```diff
@@ -635,8 +700,32 @@ func cmdAdd(args *skel.CmdArgs) error {
 		os.Setenv("CNI_IFNAME", args.IfName)
 	}
 
-	if err = setupVF(n, n.IF0, args.IfName, args.ContainerID, netns); err != nil {
-		return fmt.Errorf("failed to set up pod interface %q from the device %q: %v", args.IfName, n.IF0, err)
+	interfaceIdx, err := getInterfaceIdx(n.CNIDir)
```

Comment:
So I understand you get the PF index for the file and try to set the VF and if it failed you increase the index, but I am not sure about this approach.
What about the following:
you can loop on all the PF and check if it have free VF on the host 
like by running this "ls  /sys/class/net/enp2s0f0/device/virtfn*/net/ | grep -v -Z /sys/class | grep -v -e '^$' | wc -l" 
by the way (instead of using cmd it will be better if you can write a method that ls on each virtfnX and check if it has in the net folder interface name if so we have a free VF. 
see example below

[root@r-dcs72 r-dcs72]# ls  /sys/class/net/enp2s0f0/device/virtfn*/net/ | grep -v -Z /sys/class | grep -v -e '^$' | wc -l
8
[root@r-dcs72 r-dcs72]# #ls  /sys/class/net/enp2s0f0/device/virtfn*/net/ | grep -v -Z /sys/class | grep -v -e '^$' | wc -l
[root@r-dcs72 r-dcs72]# ip netns add test
[root@r-dcs72 r-dcs72]# ip link set enp2s0f2 netns test
[root@r-dcs72 r-dcs72]# ls  /sys/class/net/enp2s0f0/device/virtfn*/net/ | grep -v -Z /sys/class | grep -v -e '^$' | wc -l
7
[root@r-dcs72 r-dcs72]# ip netns del test
[root@r-dcs72 r-dcs72]# ls  /sys/class/net/enp2s0f0/device/virtfn*/net/ | grep -v -Z /sys/class | grep -v -e '^$' | wc -l
8
if you have free VF use this PF if not move to the next PF.

Reply:
Thx for reply. I was think about check all PF every time, but I'm afraid of creating all VF on one PF device until it has no free VF. So I decide to change PF device every time.

Reply:
@moshe010 It’s more of an approach of load balancing of available physical bandwidth, so moving to a different PF only when we exhaust the existing VFs are undesirable.

Reply:
@moshe010 I understand what are you worry about, I'll try to use sysfs only.

Reply:
@moshe010 just update the code. check the PF devices as you said, in function `getOrderedPF`
1. check all PF, and get the number of their available VF, and put them in an array
2. sort the array by number of their available VF in descending order
3. return the PF's name according to the sorted array
The `cmdAdd` will use result returned by `getOrderedPF`, to setup VF.
And I don't know whthter should I keep the `filemutex` to prevent from too many request recevied at the same time. Any suggestion?